### PR TITLE
[GEOT-6007] Build on Windows not working due to failing tests

### PR DESCRIPTION
**Pull request by vjeetje**
Fixed build For Windows users.

Current Jenkins build is failing for a while now:
http://winbuild.geo-solutions.it/jenkins/view/GeoTools/job/GeoTools-Master/

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -1530,7 +1524,7 @@
...
           </excludes>
-          <argLine>-Xmx${test.maxHeapSize} ${jvm.opts} -Dorg.geotools.test.extensive=${extensive.tests} -Dorg.geotools.test.interactive=${interactive.tests} -Dorg.geotools.image.test.skip=${skip.image.tests} -Dorg.geotools.image.test.interactive=${interactive.image} -Djava.awt.headless=${java.awt.headless} -Dsun.java2d.d3d=${sun.java2d.d3d} -Djava.io.tmpdir="${java.io.tmpdir}" -Djava.library.path="${java.library.path}" ${test.args}  -XX:+TieredCompilation -XX:TieredStopAtLevel=1 -XX:+IgnoreUnrecognizedVMOptions --add-modules=java.xml.bind --add-modules=java.activation</argLine>
+          <argLine>-Xmx${test.maxHeapSize} ${jvm.opts} -Dorg.geotools.test.extensive=${extensive.tests} -Dorg.geotools.test.interactive=${interactive.tests} -Dorg.geotools.image.test.skip=${skip.image.tests} -Dorg.geotools.image.test.interactive=${interactive.image} -Djava.awt.headless=${java.awt.headless} -Dsun.java2d.d3d=${sun.java2d.d3d} -Djava.io.tmpdir=${java.io.tmpdir} -Djava.library.path="${java.library.path}" ${test.args}  -XX:+TieredCompilation -XX:TieredStopAtLevel=1 -XX:+IgnoreUnrecognizedVMOptions --add-modules=java.xml.bind --add-modules=java.activation</argLine>
```
Comment (aaime) [inline pom.xml]: Why the change here?

File: modules/plugin/coverage-multidim/netcdf/src/test/java/org/geotools/coverage/io/netcdf/NetCDFMosaicReaderTest.java [hunk 2]
```diff
@@ -231,7 +231,7 @@ public void testHeterogeneous()
         File nc1 = TestData.file(this, "polyphemus_20130301_test.nc");
         File mosaic = new File(TestData.file(this, "."), "nc_poly_hetero");
         if (mosaic.exists()) {
-            FileUtils.deleteDirectory(mosaic);
+            FileUtils.forceDelete(mosaic);
```
Comment (aaime) [inline modules/plugin/coverage-multidim/netcdf/src/test/java/org/geotools/coverage/io/netcdf/NetCDFMosaicReaderTest.java]: So all these changes from deleteDirectory to forceDelete, are they there to ensure an exception is thrown in case the deletion does not work? Or... something else?

File: modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/footprint/SidecarFootprintProvider.java [hunk 3]
```diff
@@ -219,7 +219,7 @@ private String getFullPath(String strValue) throws IOException {
         if (!file.isAbsolute()) {
             file = new File(reference, strValue);
         }
-        return file.getCanonicalPath();
+        return file.getPath();
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/footprint/SidecarFootprintProvider.java]: Can you explain this change?

File: pom.xml (hunk 1 above)
Comment (vjeetje) [reply pom.xml]: This is to give better support for people who a space in their directory structure towards their java library / temp directory. It gave me a lot of troubles for running the tests in Intellij.

File: modules/plugin/coverage-multidim/netcdf/src/test/java/org/geotools/coverage/io/netcdf/NetCDFMosaicReaderTest.java (hunk 2 above)
Comment (vjeetje) [reply modules/plugin/coverage-multidim/netcdf/src/test/java/org/geotools/coverage/io/netcdf/NetCDFMosaicReaderTest.java]: During the running of the tests I noticed they were sometimes struggling with the deleteDirectory. I'm not sure why tbh, I guess some other tests are influencing this one. I found it hard to locate the tests that were influencing this one, so I tried forceDeleting it. That solved the problem for me. It's a recurring theme in the tests that they sometimes don't cleanup properly.

File: modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/footprint/SidecarFootprintProvider.java (hunk 3 above)
Comment (vjeetje) [reply modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/footprint/SidecarFootprintProvider.java]: There is a test that struggles on Windows with constructing the full path. It was a mosaic test, I forgot the exact name. On line 220 in that test the reference was a file path to the target folder of the test e.g. c/windows/.../target/ and the strValue was something like http://localhost:9080/mosaic. So the file path became: c/windows/.../target/http://localhost:9080/mosaic which doesn't make any sense. The test data that was given, seemed ok at the right places. The file path it constructed wasn't really a valid path. When java tries to make it canonical (standardise) it, it breaks. I see no reason why someone would want to get the canonical version of the path. The only reason one would want to do that, is to compare file paths.

File: pom.xml (hunk 1 above)
Comment (aaime) [reply pom.xml]: @vjeetje I don't understand, with paths with spaces the double quotes should be preserved, not removed  "tmpdir" as you did above?

**Comment by aaime** [comment]
I have tested the build on Linux in a path without spaces, seems to work fine.
@dromagnoli could you give it a try on Windows?
@bencaradocdavies you have a "linux build with spaces in path" scenario right? Any chance you could give this a kick too?

**Comment by bencaradocdavies** [comment]
@aaime my build is in a path with spaces but not my tmpdir or java installation. I just built GeoTools with `-Djava.io.tmpdir="/tmp/geotools with spaces"` as a test and it built fine *without* this pull request. I do not think the quotes around the `argLine`  `tmpdir` expansion should be removed. @vjeetje I do not understand why you want to make these `argLine` changes.

File: pom.xml (hunk 1 above)
Comment (bencaradocdavies) [inline pom.xml]: @vjeetje please explain or revert the `argLine` `tmpdir` quote removal.

File: modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/footprint/SidecarFootprintProvider.java (hunk 3 above)
Comment (bencaradocdavies) [reply modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/footprint/SidecarFootprintProvider.java]: @vjeetje this must be a bug in the code or a test. Nothing should be calling this method with the string encoding of a URL. @aaime is this a bug in `SidecarFootprintProvider`? Can the `location` be an HTTP URL?

**Comment by bencaradocdavies** [comment]
It would also be nice if the `dispose` calls were in `finally` blocks. It would be even nicer if `ImageReader` implemented `AutoCloseable`, then you could use try-with-resources.

File: modules/library/metadata/src/test/java/org/geotools/util/URLsTest.java [hunk 4]
```diff
@@ -101,17 +101,17 @@ public void testUrlToFile() throws Exception {
         assertURL(replaceSlashes("C:\\one"), "file://C:/one");
         assertURL(replaceSlashes("C:\\one\\two"), "file://C:/one/two");
         assertURL(replaceSlashes("C:\\one\\two\\and three"), "file://C:/one/two/and three");
-        assertEquals("sample", URLs.urlToFile(new URL("file:sample?query")).toString());
-        assertEquals("sample", URLs.urlToFile(new URL("file:sample#ref")).toString());
+        assertURL("sample", "file:sample?query");
```
Comment (bencaradocdavies) [inline modules/library/metadata/src/test/java/org/geotools/util/URLsTest.java]: @vjeetje these `assertEquals` to `assertURL` changes are good. I do not know how we missed these. Thanks!

File: modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/footprint/SidecarFootprintProvider.java (hunk 3 above)
Comment (aaime) [reply modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/footprint/SidecarFootprintProvider.java]: @bencaradocdavies before Niels changes to allow mosaic pulling from random URL definitely not, now I'm not so sure.

**Comment by aaime** [comment]
@bencaradocdavies I agree on AutoCloseable being the way to go, but don't think it's on @vjeetje plate to do so, it's an API change that will require a GSIP (switching from "dispose" to "close", deprecate the older, fix all calling points). If you're interested maybe we can do it as a team and split the work (in a separate pull request)?

**Comment by bencaradocdavies** [comment]
@aaime agreed that `AutoCloseable` is out of scope for this PR. There are many interfaces that should be considered; I will email the lists to gauge interest and scope.

File: modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/footprint/SidecarFootprintProvider.java (hunk 3 above)
Comment (vjeetje) [reply modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/footprint/SidecarFootprintProvider.java]: @aaime @bencaradocdavies The test file that is containing the url which was parsed is: org/geotools/gce/imagemosaic/test-data/remote_test/remote_test.dbf . It contains the remote url  "http://localhost:8089/global_mosaic_0.png". Is this a stable feature that somehow works on linux?

Comment (aaime) [reply modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/footprint/SidecarFootprintProvider.java]: That was recently introduced in https://github.com/geotools/geotools/pull/1869 and already caused another regression fixed in https://github.com/geotools/geotools/pull/1910
It does not break the build on Linux, but don't believe it's actually working either, that code should probably have been amended to also work with URLs instead of files. I'm however not sure why it gets into the footprint management path...

File: pom.xml (hunk 1 above)
Comment (vjeetje) [reply pom.xml]: @bencaradocdavies This is some weird Intellij interaction with maven-surefire-plugin. I'm not sure about the cause but running a test on Intellij with -Djava.io.tmpdir="${java.io.tmpdir}" -Djava.library.path="${java.library.path}" fails with a parsing problem. Removing the quotes around tmpdir fixes those issues. The build in general works fine though without my changes. I'll revert those changes as I can't find a solution that works in general on all systems with the IDE.  I've added a screenshot showing the work-around for those on Windows using Intellij. Just uncheck the argline checkbox.
argline_intellij

Comment (vjeetje) [reply pom.xml]: @aaime see my reply at the code change request of @bencaradocdavies

**Comment by vjeetje** [comment]
I've added my changes in a new commit:

- Revert the ArgLine quoting for the maven-surefire-plugin. See change request @bencaradocdavies 
- I've fixed some tests which were broken again. The cause could be that the tests are flaky or some new changes introduced some issues on Windows again.

This reminds me; how can we assure (once this pull request is accepted) that the Windows build will keep succeeding? Can we configure Travis / enforce some rule?

**Comment by bencaradocdavies** [comment]
@vjeetje GeoSolutions had a Windows Jenkins instance but it has been muted(?) for some time because the build was failing. Once the Windows build is working, and we have Windows developers available to fix Windows-only failures, Windows build notifications could be re-enabled. @aaime asked @dromagnoli to try building the PR on Windows, which suggests that we may have someone with a Windows development environment.

I only use Linux but always build in a path with spaces to catch File<->URL conversion problems, which are a common source of failures on Windows and non-ASCII platforms.

**Comment by aaime** [comment]
@vjeetje Ben is correct, we'll need further help to keep it going.
I hate working on windows, but I was the one to push for setting up that build server and get the build going by crushing a few days off in a Christmas break on the accursed OS ;-)
I did so because multiple people complained about the build not working on Windows, so was trying to be a good citizen.
Unfortunately all complainers disappeared once the build was fixed and the build went back to broken state within a relatively short time (a few months I think).
So, the build server is there, we can re-enable hate mails from it, but we'll also need some support from someone that actually cares to keep the Windows build going (as in, cares about that more than about other things to do in this community, for which we have plenty).

**Comment by vjeetje** [comment]
@aaime Maybe checking the Windows build should be part of the code reviewing process.

**Comment by aaime** [comment]
@vjeetje there are two issues:
* We don't have a windows build for pull requests
* Pull requests are not required for direct project committers (the vast majority of are not using windows), which normally will use PRs only if they think a review is needed, or if they feel like having a second opinion from Travis

For the first one, are you interested in doing research and propose an approach? 
Believe there are free build server that could be paired with PR checks, but never looked into it.

For the second one, I personally would not want to go through PRs for every commit, would slow me down too much, but if a majority or the other devs decide it's a good idea, I'll eat crow.

**Comment by aaime** [comment]
@vjeetje both are a no go, mostly because you're using the wrong pronoun ;-)
"We" should setup AppVeyour (on our own spare time, nobody is sponsoring windows support) and "we" should do PR and review for every change (same as above, nobody is authorized to use extra working hours for this).

The main problem is that "we" are already way overbooked, and are cutting dead branches on whatever we cannot maintain with extra effort (we are actually in the process dropping modules down to unsupported to reduce the load).

It really boils down to was "you" can do, as a new contributor, to help with the Windows build. Otherwise I'm skeptical anything of significance will happen.

**Comment by aaime** [comment]
@bencaradocdavies aside from the other discussions, are you happy with the changes? Is the pull request mergeable now?

**Comment by NielsCharlier** [comment]
It seems indeed that I might have missed another URL-to-file conversion, for example
https://github.com/geotools/geotools/blob/master/modules/plugin/imagemosaic/src/main/java/org/geotools/gce/imagemosaic/ImageMosaicConfigHandler.java#L946

None of my test broke on this though, not sure why

**Comment by aaime** [comment]
@NielsCharlier see the one in this pull request, looks like on Linux the conversion is lenient, on Windows is not. Probably just one of the possible causes.

**Comment by dromagnoli** [comment]
@aaime I successfully did the build on Windows... no issues there

File: modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/footprint/SidecarFootprintProvider.java (hunk 3 above)
Comment (bencaradocdavies) [inline modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/footprint/SidecarFootprintProvider.java]: I think that the `file.getCanonicalPath()` to `file.getPath()` change in `SidecarFootprintProvider.getFullPath` is the wrong fix because this calling code should not be passing the string representation of an HTTP(S) URL in the first place. I would prefer a guard in the calling code to prevent HTTP(S) URLs from being used until they are supported. For example, change line 117 to:
```
            if (value != null && value instanceof String
                    && !((String) value).matches("https?://.*")) {
```

Comment (vjeetje) [reply modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/footprint/SidecarFootprintProvider.java]: Done

**Comment by bencaradocdavies** [comment]
Thanks @vjeetje, that looks good. I also note your improvements to the guard regex.
